### PR TITLE
8263614: javac allows local variables to be accessed from a static context

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
javac breaks with NPE if compiles this code:

```
public class LocalClasses {
    public static void main(String[] args) {
        int i = 5;
        class Local {
            static void m() {
                System.out.println("Value of i = " + i);
            }
        }
        Local.m();
    }
} 
```

actually the compiler should issue a compiling error as local variable `i` is being accessed from a static context. Please review this fix to address this issue. Some notes:

`com.sun.tools.javac.comp.AttrContext.staticLevel` keeps a the number of nested static contexts but this calculation doesn't consider static type declarations. This is because static declaration doesn't introduce a static context. But if they have a static modifier, even if implicit as it is for local records, then this affects what variables, type variables, etc are accessible from the body of the static type declaration. For this reason I have used an `adjusted` static level that takes static type declarations into account.

TIA

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263614](https://bugs.openjdk.java.net/browse/JDK-8263614): javac allows local variables to be accessed from a static context

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4004/head:pull/4004` \
`$ git checkout pull/4004`

Update a local copy of the PR: \
`$ git checkout pull/4004` \
`$ git pull https://git.openjdk.java.net/jdk pull/4004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4004`

View PR using the GUI difftool: \
`$ git pr show -t 4004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4004.diff">https://git.openjdk.java.net/jdk/pull/4004.diff</a>

</details>

### CODE REVIEW

**Comment by mcimadamore** [comment]
This is a tricky area, so I played a little with the (unchanged) code, to get an idea of what was wrong. It seems to me that the behavior of `Resolve::findVar` (other methods are probably the same) is as follows:

0. set `staticOnly` to `true` is the current env is static
1. look in local variable of current env
2. look in fields of current env
3. if a field is found and the field is compatible with `staticOnly` return that, otherwise issue a static error
4. go back to (1) with the enclosing env, setting `staticOnly` if a static member is found

This doesn't seem *too* bad - but there is an issue, e.g. in the example mentioned in the JBS issue: we look into locals (step 1) regardless of the value of `isStatic`. This leads to issue, because it doesn't make sense to "find" a local variable in an enclosing scope if the type from which the search started is a static type.

To rectify this issue, I came up with the following, relatively minimal fix:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
index 3d4859d1b88..cdc129e8ad6 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
@@ -1510,13 +1510,16 @@ public class Resolve {
         boolean staticOnly = false;
         while (env1.outer != null) {
             Symbol sym = null;
-            if (isStatic(env1)) staticOnly = true;
             for (Symbol s : env1.info.scope.getSymbolsByName(name)) {
                 if (s.kind == VAR && (s.flags_field & SYNTHETIC) == 0) {
                     sym = s;
+                    if (staticOnly) {
+                        return new StaticError(sym);
+                    }
                     break;
                 }
             }
+            if (isStatic(env1)) staticOnly = true;
             if (sym == null) {
                 sym = findField(env1, env1.enclClass.sym.type, name, env1.enclClass.sym);
             }
```

The idea here is that, whatever symbol is found during step (1), if `staticOnly` is set, we should just disregard it and throw a static error instead (as done for the field lookup in (3). This seems to fix the issue in JBS, and seems to pass all javac test.

It is likely that, as in your patch, similar changes would be needed for other routines such as `Resolve::findType`/`findTypeVar`.

**Comment by vicente-romero-oracle** [comment]
thanks Maurizio for taking the time to dig deeper into this issue. Your patch is elegant but we will need some additional changes if we decide to go through that path, for example the compiler is accepting this code with your patch instead of issuing an error:

```
class C<T> {
    int field = 0;
    <U> void foo(int param) {
        int localVar = 1;
        class Local {
            static void m() {
                U u;
            }
        }
    }
}
```

**Comment by vicente-romero-oracle** [comment]
@mcimadamore, I have uploaded another commit which builds on your patch and adds a small change to Resolve::findTypeVar, thanks

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 1]
```diff
@@ -2323,9 +2326,11 @@ Symbol findTypeVar(Env<AttrContext> currentEnv, Env<AttrContext> originalEnv, Na
                     allowRecords &&
                     (sym.owner.kind == MTH &&
                     currentEnv != originalEnv &&
-                    !isInnerClassOfMethod(sym.owner, originalEnv.tree.hasTag(CLASSDEF) ?
-                            ((JCClassDecl)originalEnv.tree).sym :
-                            originalEnv.enclClass.sym)))) {
+                        (!isInnerClassOfMethod(sym.owner, originalEnv.tree.hasTag(CLASSDEF) ?
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:2329]: Can't you do a similar approach here and use staticOnly to rule out tvars?

Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:2329]: I think you can go in findType, move the assignment to staticOnly to *after* the call to findTypeVar - then you can drop all type var symbol you find with staticOnly = true.

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:2329]: I think that type variables are trickier than local variables and probably won't be as easy to tame. I tried several combinations of this patch which I think is what you have in mind:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
index 70ea875a4b3..832a8ba8ede 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
@@ -2316,29 +2316,6 @@ public class Resolve {
         return bestSoFar;
     }
 
-    Symbol findTypeVar(Env<AttrContext> currentEnv, Env<AttrContext> originalEnv, Name name, boolean staticOnly) {
-        for (Symbol sym : currentEnv.info.scope.getSymbolsByName(name)) {
-            if (sym.kind == TYP) {
-                if (staticOnly &&
-                    sym.type.hasTag(TYPEVAR) &&
-                    ((sym.owner.kind == TYP) ||
-                    // are we trying to access a TypeVar defined in a method from a local static type: interface, enum or record?
-                    allowRecords &&
-                    (sym.owner.kind == MTH &&
-                    currentEnv != originalEnv &&
-                        (!isInnerClassOfMethod(sym.owner, originalEnv.tree.hasTag(CLASSDEF) ?
-                        ((JCClassDecl)originalEnv.tree).sym :
-                        originalEnv.enclClass.sym) ||
-                        originalEnv.info.staticLevel > currentEnv.info.staticLevel)
-                    ))) {
-                    return new StaticError(sym);
-                }
-                return sym;
-            }
-        }
-        return typeNotFound;
-    }
-
     boolean isInnerClassOfMethod(Symbol msym, Symbol csym) {
         while (csym.owner != msym) {
             if (csym.isStatic()) return false;
@@ -2358,18 +2335,25 @@ public class Resolve {
         Symbol sym;
         boolean staticOnly = false;
         for (Env<AttrContext> env1 = env; env1.outer != null; env1 = env1.outer) {
-            if (isStatic(env1)) staticOnly = true;
             // First, look for a type variable and the first member type
-            final Symbol tyvar = findTypeVar(env1, env, name, staticOnly);
+            Symbol tyvar = typeNotFound;
+            for (Symbol s : env1.info.scope.getSymbolsByName(name)) {
+                if (s.kind == TYP) {
+                    tyvar = s;
+                }
+            }
             sym = findImmediateMemberType(env1, env1.enclClass.sym.type,
                                           name, env1.enclClass.sym);
-
+            if (isStatic(env1)) staticOnly = true;
             // Return the type variable if we have it, and have no
             // immediate member, OR the type variable is for a method.
             if (tyvar != typeNotFound) {
                 if (env.baseClause || sym == typeNotFound ||
                     (tyvar.kind == TYP && tyvar.exists() &&
                      tyvar.owner.kind == MTH)) {
+                    if (staticOnly) {
+                        return new StaticError(tyvar);
+                    }
                     return tyvar;
                 }
             }
```

**Comment by mcimadamore** [comment]
I'm doing more investigation. The code for resolving types is definitively complex. It seems to me that the complexity comes from 3 problems:

Problem 1:

```
class Foo<T> {
static T t;
}
```

Here the env for `t` has ` scope which can see T.

Problem 2:

```
class Foo<T> {
static void m() {
   T t;
}
}
```

Here the env for `m` has ` scope which can see T.

Problem 3:

```
class Foo {
static <Z> void m() {
   Z z;
}
}
```

This should obviously work, although, in compiler terms, there's not much difference between (3) and (2).

I'm currently looking into all the changes made in this area in recent times, and see if there's an easier path to support all this.

**Comment by mcimadamore** [comment]
This is what I came up with:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
index 86f68f94efd..a325a08459f 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
@@ -1488,30 +1488,24 @@ public class Resolve {
         boolean staticOnly = false;
         while (env1.outer != null) {
             Symbol sym = null;
-            if (isStatic(env1)) staticOnly = true;
             for (Symbol s : env1.info.scope.getSymbolsByName(name)) {
                 if (s.kind == VAR && (s.flags_field & SYNTHETIC) == 0) {
                     sym = s;
+                    if (staticOnly) {
+                        return new StaticError(sym);
+                    }
                     break;
                 }
             }
+            if (isStatic(env1)) staticOnly = true;
             if (sym == null) {
                 sym = findField(env1, env1.enclClass.sym.type, name, env1.enclClass.sym);
             }
             if (sym.exists()) {
                 if (staticOnly &&
-                   (sym.flags() & STATIC) == 0 &&
-                    sym.kind == VAR &&
-                        // if it is a field
-                        (sym.owner.kind == TYP ||
-                        // or it is a local variable but it is not declared inside of the static local type
-                        // then error
-                        allowRecords &&
-                        (sym.owner.kind == MTH) &&
-                        env1 != env &&
-                        !isInnerClassOfMethod(sym.owner, env.tree.hasTag(CLASSDEF) ?
-                                ((JCClassDecl)env.tree).sym :
-                                env.enclClass.sym)))
+                        sym.kind == VAR &&
+                        sym.owner.kind == TYP &&
+                        (sym.flags() & STATIC) == 0)
                     return new StaticError(sym);
                 else
                     return sym;
@@ -2313,35 +2307,22 @@ public class Resolve {
         return bestSoFar;
     }
 
-    Symbol findTypeVar(Env<AttrContext> currentEnv, Env<AttrContext> originalEnv, Name name, boolean staticOnly) {
-        for (Symbol sym : currentEnv.info.scope.getSymbolsByName(name)) {
+    Symbol findTypeVar(Env<AttrContext> env, Name name, boolean staticOnly) {
+        for (Symbol sym : env.info.scope.getSymbolsByName(name)) {
             if (sym.kind == TYP) {
-                if (staticOnly &&
-                    sym.type.hasTag(TYPEVAR) &&
-                    ((sym.owner.kind == TYP) ||
-                    // are we trying to access a TypeVar defined in a method from a local static type: interface, enum or record?
-                    allowRecords &&
-                    (sym.owner.kind == MTH &&
-                    currentEnv != originalEnv &&
-                    !isInnerClassOfMethod(sym.owner, originalEnv.tree.hasTag(CLASSDEF) ?
-                            ((JCClassDecl)originalEnv.tree).sym :
-                            originalEnv.enclClass.sym)))) {
+                if (sym.type.hasTag(TYPEVAR) &&
+                        (staticOnly || (isStatic(env) && sym.owner.kind == TYP)))
+                    // if staticOnly is set, it means that we have recursed through a static declaration,
+                    // so type variable symbols should not be accessible. If staticOnly is unset, but
+                    // we are in a static declaration (field or method), we should not allow type-variables
+                    // defined in the enclosing class to "leak" into this context.
                     return new StaticError(sym);
-                }
                 return sym;
             }
         }
         return typeNotFound;
     }
 
-    boolean isInnerClassOfMethod(Symbol msym, Symbol csym) {
-        while (csym.owner != msym) {
-            if (csym.isStatic()) return false;
-            csym = csym.owner.enclClass();
-        }
-        return (csym.owner == msym && !csym.isStatic());
-    }
-
     /** Find an unqualified type symbol.
      *  @param env       The current environment.
      *  @param name      The type's name.
@@ -2353,9 +2334,9 @@ public class Resolve {
         Symbol sym;
         boolean staticOnly = false;
         for (Env<AttrContext> env1 = env; env1.outer != null; env1 = env1.outer) {
-            if (isStatic(env1)) staticOnly = true;
             // First, look for a type variable and the first member type
-            final Symbol tyvar = findTypeVar(env1, env, name, staticOnly);
+            final Symbol tyvar = findTypeVar(env1, name, staticOnly);
+            if (isStatic(env1)) staticOnly = true;
             sym = findImmediateMemberType(env1, env1.enclClass.sym.type,
                                           name, env1.enclClass.sym);
 ```
 
 It reverts some of the code changes made for records, and reverts the code to a simpler state (the isInnerClassOfMethod) is no longer needed. This passes all tests, including the modified RecordCompilationTests.

**Review by mcimadamore — APPROVED** [review]
Looks good

**Pull request closed without merging**